### ISSUE TITLE
Complete event is not called if the /json endpoint results in an error

### ISSUE THREAD

**Issue by jwrosewell** [51Degrees/device-detection-php-onpremise#12]
Issue #11 relates to the example not responding to the end point because the settings/configuration is not being set as expected. The impact is that the `fod.Complete` event does not fire.

However there are other reasons that would prevent the `fod.Complete` from firing. For example; the CORS configuration is not correct if the JavaScript and JSON endpoints on are different origins to the calling web page.

We need to review the error handling associated with the subsequent calls to the JSON end point and present clearer error messages.

**Comment by justadreamer** [comment]
Also the deprecation warnings output by PHP are preventing `fod.complete` being called, even if #11 is fixed - the response produced by the `/json` endpoint outputs some warnings and it is no longer a valid JSON :( - so the code will fail somewhat earlier than calling `fod.complete`.  

```json
<br />
<b>Deprecated</b>:  Creation of dynamic property fiftyone\pipeline\devicedetection\DeviceDetectionOnPremise::$engine is deprecated in <b>/Users/eugene/projects/51degrees/51drepos/device-detection-php-onpremise/DeviceDetectionOnPremise.php</b> on line <b>46</b><br />
<br />
<b>Deprecated</b>:  Creation of dynamic property fiftyone\pipeline\devicedetection\DeviceDetectionOnPremise::$evidenceKeys is deprecated in <b>/Users/eugene/projects/51degrees/51drepos/device-detection-php-onpremise/DeviceDetectionOnPremise.php</b> on line <b>91</b><br />
<br />
<b>Deprecated</b>:  Creation of dynamic property fiftyone\pipeline\devicedetection\SwigData::$result is deprecated in <b>/Users/eugene/projects/51degrees/51drepos/device-detection-php-onpremise/SwigData.php</b> on line <b>36</b><br />
{"javascriptProperties":["device.javascripthardwareprofile","device.javascriptbandwidth","device.javascriptimageoptimiser","device.screenpixelsheightjavascript","device.screenpixelswidthjavascript","device.pixelratiojavascript","device.javascriptgethighentropyvalues"],"device":{"ismobile":false,"ismobilenullreason":"Unknown","istablet":false,"istabletnullreason":"Unknown","screenpixelswidth":1920,"screenpixelsheight":1080,"hastouchscreen":false,"hastouchscreennullreason":"Unknown","hasqwertypad":true,"hardwarevendor":"Apple","hardwaremodel":"Macintosh","hardwarename":["Macintosh"],"isconsole":false,"isconsolenullreason":"Unknown","platformvendor":"Apple","platformname":"macOS","platformversion":"13.5","layoutengine":"Unknown","browservendor":"Microsoft","browsername":"Edge (Chromium) for Macintosh","screenmmwidth":0,"screenmmwidthnullreason":"Unknown","screenmmheight":0,"screenmmheightnullreason":"Unknown","bitsperpixel":0,"bitsperpixelnullreason":"Unknown","browserversion":"115","cpu":"Unknown","ccppaccept":["application\/java","application\/java-archive","application\/vnd.oma.dd+xml","application\/vnd.oma.drm.content","application\/vnd.oma.drm.message","application\/vnd.oma.drm.rights+wbxml","application\/vnd.oma.drm.rights+xml","application\/vnd.wap.multipart.mixed","application\/vnd.wap.wbxml","application\/vnd.wap.wmlscriptc","application\/vnd.wap.xhtml+xml","application\/x-compress","application\/x-gzip","application\/xhtml+xml","audio\/3ga","audio\/3gpp","audio\/MP4A-ES","audio\/MP4A-LATM","audio\/aac","audio\/amr","audio\/asf","audio\/asx","audio\/awb","audio\/m4a","audio\/m4b","audio\/midi","audio\/mp3","audio\/mp3d","audio\/mp4","audio\/wav","audio\/wma","image\/2bp","image\/bmp","image\/gif","image\/jpeg","image\/jpg","image\/png","image\/tif","image\/tiff","image\/vnd.wap.wbmp","image\/wbmp","message\/rfc822","text\/css","text\/html","text\/plain","text\/vnd.sun.j2me.app-descriptor","text\/vnd.wap.wml","text\/x-ical","text\/x-vcal","text\/xml","text\/xsl","video\/3g2","video\/3gp","video\/3gpp","video\/H.263-2000","video\/H.264","video\/MP4V-ES","video\/divx","video\/flv","video\/m4v","video\/mp4","video\/wmv","video\/xvid"],"htmlversion":0,"htmlversionnullreason":"Unknown","javascript":true,"javascriptversion":"1.7","supportedbearers":["Unknown"],"supportedbluetoothversion":"Unknown","cpumaximumfrequency":0,"cpumaximumfrequencynullreason":"Unknown","releaseyear":0,"releaseyearnullreason":"Unknown","releasemonth":"Unknown","cookiescapable":true,"streamingaccept":["Unknown"],"hasvirtualqwerty":false,"hasvirtualqwertynullreason":"Unknown","haskeypad":false,"haskeypadnullreason":"Unknown","hascamera":false,"hascameranullreason":"Unknown","backcameramegapixels":0,"backcameramegapixelsnullreason":"Unknown","expansionslottype":["Unknown"],"expansionslotmaxsize":0,"expansionslotmaxsizenullreason":"Unknown","deviceram":0,"deviceramnullreason":"Unknown","maxinternalstorage":0,"maxinternalstoragenullreason":"Unknown","suggestedimagebuttonheightpixels":11.8,"suggestedimagebuttonheightmms":3.5,"suggestedlinksizepixels":11.8,"suggestedlinksizepoints":10,"hastrackpad":true,"hasclickwheel":false,"hasclickwheelnullreason":"Unknown","isereader":false,"isereadernullreason":"Unknown","javascriptcanmanipulatedom":true,"javascriptcanmanipulatecss":true,"javascriptsupportseventlistener":false,"javascriptsupportseventlistenernullreason":"Unknown","javascriptsupportsevents":true,"javascriptgetelementbyid":true,"ajaxrequesttype":"Unknown","javascriptsupportsinnerhtml":true,"javascriptpreferredgeolocapi":"W3C api","screenmmdiagonal":0,"screenmmdiagonalnullreason":"Unknown","html5video":["H.264","Ogg (Theora)","WebM"],"html5audio":["AAC","MP3","Wav"],"csscolumn":false,"csscolumnnullreason":"Unknown","csstransitions":false,"csstransitionsnullreason":"Unknown","has3dcamera":false,"has3dcameranullreason":"Unknown","has3dscreen":false,"has3dscreennullreason":"Unknown","iscrawler":false,"iscrawlernullreason":"Unknown","crawlername":"NotCrawler","jquerymobilesupport":"Unknown","cameratypes":["Unknown"],"cpucores":0,"cpucoresnullreason":"Unknown","animationtiming":false,"animationtimingnullreason":"Unknown","blobbuilder":false,"blobbuildernullreason":"Unknown","cssbackground":false,"cssbackgroundnullreason":"Unknown","cssborderimage":false,"cssborderimagenullreason":"Unknown","csscanvas":false,"csscanvasnullreason":"Unknown","csscolor":false,"csscolornullreason":"Unknown","cssflexbox":false,"cssflexboxnullreason":"Unknown","cssfont":false,"cssfontnullreason":"Unknown","cssimages":false,"cssimagesnullreason":"Unknown","cssmediaqueries":false,"cssmediaqueriesnullreason":"Unknown","cssminmax":false,"cssminmaxnullreason":"Unknown","cssoverflow":false,"cssoverflownullreason":"Unknown","cssposition":false,"csspositionnullreason":"Unknown","csstext":false,"csstextnullreason":"Unknown","csstransforms":false,"csstransformsnullreason":"Unknown","cssui":false,"cssuinullreason":"Unknown","dataset":false,"datasetnullreason":"Unknown","dataurl":false,"dataurlnullreason":"Unknown","deviceorientation":false,"deviceorientationnullreason":"Unknown","filereader":false,"filereadernullreason":"Unknown","filesaver":false,"filesavernullreason":"Unknown","filewriter":false,"filewriternullreason":"Unknown","formdata":false,"formdatanullreason":"Unknown","fullscreen":false,"fullscreennullreason":"Unknown","geolocation":false,"geolocationnullreason":"Unknown","history":false,"historynullreason":"Unknown","html-media-capture":false,"html-media-capturenullreason":"Unknown","html5":true,"iframe":false,"iframenullreason":"Unknown","indexeddb":false,"indexeddbnullreason":"Unknown","json":false,"jsonnullreason":"Unknown","postmessage":false,"postmessagenullreason":"Unknown","progress":false,"progressnullreason":"Unknown","prompts":false,"promptsnullreason":"Unknown","selector":false,"selectornullreason":"Unknown","svg":false,"svgnullreason":"Unknown","touchevents":false,"toucheventsnullreason":"Unknown","track":false,"tracknullreason":"Unknown","video":false,"videonullreason":"Unknown","viewport":false,"viewportnullreason":"Unknown","webworkers":false,"webworkersnullreason":"Unknown","xhr2":false,"xhr2nullreason":"Unknown","masking":false,"maskingnullreason":"Unknown","canvas":false,"canvasnullreason":"Unknown","supportstls\/ssl":false,"supportstls\/sslnullreason":"Unknown","cldc":0,"cldcnullreason":"Unknown","midp":0,"midpnullreason":"Unknown","screeninchesdiagonal":0,"screeninchesdiagonalnullreason":"Unknown","istv":false,"istvnullreason":"Unknown","screeninchesheight":0,"screeninchesheightnullreason":"Unknown","screenincheswidth":0,"screenincheswidthnullreason":"Unknown","issmartphone":false,"issmartphonenullreason":"Unknown","issmallscreen":false,"issmallscreennullreason":"Unknown","hardwareimages":["Front\thttp:\/\/images.51degrees.mobi\/DI\/Front_17595.png","Posed\thttp:\/\/images.51degrees.mobi\/DI\/Posed_17595.png"],"hasnfc":false,"hasnfcnullreason":"Unknown","hardwarefamily":"Macintosh","isemailbrowser":false,"isemailbrowsernullreason":"Unknown","oem":"Apple","supportedchargertypes":["Unknown"],"hasremovablebattery":false,"hasremovablebatterynullreason":"Unknown","batterycapacity":0,"batterycapacitynullreason":"Unknown","satellitenavigationtypes":["Unknown"],"supportedsensortypes":["Unknown"],"durability":["Unknown"],"maxnumberofsimcards":0,"maxnumberofsimcardsnullreason":"Unknown","supportedsimcardtypes":["Unknown"],"gpu":"Unknown","supportedcamerafeatures":["Unknown"],"maxtalktime":0,"maxtalktimenullreason":"Unknown","maxstandbytime":0,"maxstandbytimenullreason":"Unknown","dynamiccontrastratio":"N\/A","supportswidi":false,"supportswidinullreason":"Unknown","energyconsumptionperyear":0,"energyconsumptionperyearnullreason":"Unknown","supports24p":false,"supports24pnullreason":"Unknown","supportedi\/o":["Unknown"],"refreshrate":0,"refreshratenullreason":"Unknown","screentype":"Unknown","onpowerconsumption":0,"onpowerconsumptionnullreason":"Unknown","contrastratio":"N\/A","maxusagetime":0,"maxusagetimenullreason":"Unknown","cpudesigner":"Unknown","gpudesigner":"Unknown","frontcameramegapixels":0,"frontcameramegapixelsnullreason":"Unknown","soc":"Unknown","socdesigner":"Unknown","socmodel":"Unknown","ismediahub":false,"ismediahubnullreason":"Unknown","javascripthardwareprofile":"var profileIds = []\n\/*!\n * Copyright 51 Degrees Mobile Experts Limited.\n *\/\n\"use strict\";const defaultRetries=0;function getProfileId(p,m=defaultRetries){\n\/*! VERSION = 1.641563 *\/\nvar A={Version:\"1.641563\",PublishDate:\"2023-02-23T12:11:37.0793211Z\",Data:[{m:function(e){return window.screen.height*window.devicePixelRatio},n:[20,22,21,12,6,10,13,14,9,15,19,18,11,3,7,17,1,4,8,16,5,2]},{m:x,n:[23,24],v:[2732]},{m:x,n:[26,25],v:[2048]},{x:37878,m:x,n:[27,28],v:[2208]},{m:x,n:[29,30],v:[1334]},{m:x,n:[32,31],v:[1136]},{x:98246,m:n,n:[35,34,33],v:[2224]},{x:37878,m:x,n:[37,36],v:[2001]},{x:123083,m:n,n:[39,40,38],v:[2436]},{m:x,n:[41,42],v:[1800]},{x:99613,m:n,n:[44,43],v:[2688]},{x:123631,m:n,n:[46,45],v:[2388]},{x:\"MacBook Pro (2018)\",v:[2100]},{x:99612,m:n,n:[48,47],v:[1624]},{x:99612,m:n,n:[48,47],v:[1792]},{x:116517,m:v,n:[49,50,51,52],v:[2160]},{x:122591,m:n,n:[39,54,53],v:[2079]},{x:122587,m:e,n:[58,55,59,56,57],v:[2532]},{m:function(e){return window.devicePixelRatio},n:[61,60],v:[2778]},{m:x,n:[63,62],v:[2360]},{x:115474,v:[2266]},{x:122490,v:[2796]},{x:122489,v:[2556]},{x:85910,v:[\"srgb\"]},{x:123627,m:n,n:[65,66,64],v:[\"p3\"]},{x:98241,m:n,n:[69,70,72,71,68,73,67],v:[\"srgb\"]},{x:124548,m:n,n:[76,65,66,75,74],v:[\"p3\"]},{x:98238,m:n,n:[79,78,77],v:[\"srgb\"]},{x:98240,m:n,n:[81,82,80],v:[\"p3\"]},{x:98237,m:n,n:[85,84,83],v:[\"srgb\"]},{x:118945,m:n,n:[87,88,86],v:[\"p3\"]},{m:n,n:[92,85,91,93,90,89],v:[\"srgb\"]},{x:118945,m:n,n:[87,88,94],v:[\"p3\"]},{x:98246,m:v,n:[96,95],v:[\"Apple GPU\"]},{x:85905,v:[\"Apple A10X GPU\"]},{x:95561,v:[\"Apple A12 GPU\"]},{x:98240,m:n,n:[81,82,97],v:[\"p3\"]},{x:98238,m:n,n:[79,78,98],v:[\"srgb\"]},{x:123083,m:e,n:[100,101,102,99,104,105,103],v:[\"Apple GPU\"]},{x:81104,v:[\"Apple A11 GPU\"]},{x:96282,v:[\"Apple A12 GPU\"]},{x:\"MacBook Air (2019)\",v:[\"srgb\"]},{x:\"MacBook Pro (2019)\",v:[\"p3\"]},{x:99613,m:e,n:[107,106],v:[\"Apple GPU\"]},{x:92498,v:[\"Apple A12 GPU\"]},{x:123631,m:e,n:[109,111,110,108],v:[\"Apple GPU\"]},{x:93304,v:[\"Apple A12X GPU\"]},{x:99612,m:e,n:[113,112],v:[\"Apple GPU\"]},{x:92486,v:[\"Apple A12 GPU\"]},{x:99482,v:[2114570256]},{x:106975,v:[1349146759]},{x:116518,m:e,n:[115,114],v:[2206992415]},{x:116518,m:e,n:[115,114],v:[2866949877]},{x:122591,m:e,n:[119,122,116,117,118,121,123,120],v:[\"Apple GPU\"]},{x:92493,v:[\"Apple A12 GPU\"]},{x:108616,v:[105985484]},{x:116508,m:v,n:[125,124,126],v:[3403189785]},{x:116508,v:[2364051618]},{x:115461,v:[679860869]},{x:123079,v:[46273595]},{x:123080,m:e,n:[127,129,128],v:[3]},{x:123246,v:[2]},{x:122594,m:e,n:[130,132,131],v:[\"p3\"]},{x:123245,v:[\"srgb\"]},{x:123627,m:e,n:[134,133,137,135,138,136],v:[\"Apple GPU\"]},{x:85906,v:[\"Apple A10X GPU\"]},{x:93305,v:[\"Apple A12X GPU\"]},{x:98241,m:v,n:[140,143,141,142,144,139,145],v:[\"Apple GPU\"]},{x:85910,v:[\"Apple A9X GPU\"]},{x:85907,v:[\"Apple A9 GPU\"]},{x:86700,v:[\"Apple A10 GPU\"]},{x:85902,v:[\"Apple A8 GPU\"]},{x:85901,v:[\"Apple A8X GPU\"]},{x:85900,v:[\"Apple A7 GPU\"]},{x:124548,m:e,n:[137,135,138,147,146,148],v:[\"Apple GPU\"]},{x:97978,v:[\"Apple A9X GPU\"]},{x:95563,v:[\"Apple A12 GPU\"]},{x:98238,m:v,n:[150,149],v:[\"Apple GPU\"]},{x:57693,v:[\"Apple A9 GPU\"]},{x:50764,v:[\"Apple A8 GPU\"]},{x:98240,m:v,n:[151,152],v:[\"Apple GPU\"]},{x:69950,v:[\"Apple A10 GPU\"]},{x:81101,v:[\"Apple A11 GPU\"]},{x:98237,m:v,n:[154,153],v:[\"Apple GPU\"]},{x:57691,v:[\"Apple A9 GPU\"]},{x:50762,v:[\"Apple A8 GPU\"]},{x:118945,m:v,n:[155,157,160,159,156,158],v:[\"Apple GPU\"]},{x:69949,v:[\"Apple A10 GPU\"]},{x:81100,v:[\"Apple A11 GPU\"]},{x:122500,m:v,n:[164,154,162,163,161],v:[\"Apple GPU\"]},{x:85790,m:e,n:[165,166],v:[\"Apple A9 GPU\"]},{x:101895,v:[\"Apple A10 GPU\"]},{x:50755,v:[\"Apple A7 GPU\"]},{x:85788,m:v,n:[168,167],v:[\"PowerVR SGX 543\"]},{x:118945,m:e,n:[169,170,171,173,172],v:[\"Apple GPU\"]},{x:85905,v:[63583436,2114570256,3129316290,3249312110]},{x:95561,v:[1349146759,2917249763]},{x:98240,m:v,n:[151,174],v:[\"Apple GPU\"]},{x:98238,m:v,n:[150,175],v:[\"Apple GPU\"]},{x:96282,v:[958581112,2301174800,4085158452]},{x:81104,v:[367695777,411650080,1220644697]},{x:99603,v:[1278953537,3335845976,4193218782]},{x:108677,v:[105985484]},{x:123106,m:v,n:[179,176,177,178],v:[3403189785]},{x:117357,v:[2364051618]},{x:123109,v:[46273595]},{x:92498,v:[958581112,2301174800,4085158452]},{x:99455,v:[352823931,1278953537,3335845976,4193218782]},{x:113177,v:[4085158452]},{x:112384,v:[105985484,2364051618]},{x:123632,m:v,n:[180,181],v:[3403189785]},{x:123246,v:[46273595]},{x:92486,v:[958581112,2301174800,3403189785,4085158452]},{x:99442,v:[352823931,3335845976,4193218782]},{x:106975,v:[2301174800]},{x:115478,v:[3335845976]},{x:81104,v:[367695777,411650080]},{x:92493,v:[958581112,2301174800,4085158452]},{x:108704,v:[105985484]},{x:99441,v:[352823931,1278953537,3335845976]},{x:117892,v:[2364051618]},{x:123082,m:v,n:[179,182,183],v:[3403189785]},{x:107549,v:[679860869]},{x:123157,v:[46273595]},{x:108616,v:[1349146759]},{x:107548,v:[1444462398]},{x:122809,v:[2775654583]},{x:107555,v:[105985484]},{x:116511,v:[2364051618]},{x:115469,v:[1407135659]},{x:106973,v:[105985484]},{x:122594,m:v,n:[184,185],v:[3403189785]},{x:122593,v:[2364051618]},{x:93305,v:[4085158452]},{x:85906,v:[583354101,3458129248,3928876783]},{x:112379,v:[105985484,2364051618]},{x:123629,m:v,n:[187,188,189,186],v:[3403189785]},{x:103206,v:[958581112,2301174800,2487400911]},{x:123247,v:[46273595]},{x:97983,v:[63583436,2114570256,3129316290,3249312110]},{x:86700,v:[46663968]},{x:96271,m:e,n:[190,191],v:[4005673483]},{x:96271,v:[1361285941]},{x:85901,v:[1350183384,3816812018,4125234388]},{x:96271,m:e,n:[193,192],v:[2870741841]},{x:85900,v:[1915583345]},{x:97942,v:[583354101,3458129248,3928876783]},{x:98126,v:[4085158452]},{x:123628,m:v,n:[194,189],v:[3403189785]},{x:57693,v:[46663968,2114570256,3129316290]},{x:50764,v:[4125234388]},{x:69950,v:[63583436,2114570256,3129316290]},{x:81101,v:[2206992415,2917249763,2946940121,3237505312]},{x:57691,v:[46663968,63583436,2114570256,3129316290]},{x:50762,v:[4125234388]},{x:69949,v:[46663968,63583436,2114570256,3129316290]},{x:106935,m:e,n:[195,196],v:[1349146759]},{x:81100,v:[2917249763,2946940121,3237505312]},{x:106935,m:e,n:[198,197],v:[2206992415]},{x:118711,v:[1444462398,2775654583]},{x:103758,v:[2866949877]},{x:108371,v:[3129316290]},{x:62734,v:[46663968]},{x:85790,v:[63583436,2114570256]},{x:50755,v:[1915583345]},{x:57691,v:[3458129248]},{x:62734,v:[3403189785]},{x:50750,v:[330115695,887223528,1282943734,1336716766,1612709839,2403130013,2908735491,3366809836]},{x:50739,v:[155223211,1414873384,1429108241,2037784901,2156491183,3022790293]},{x:69949,v:[583354101,3458129248,3928876783]},{x:81100,v:[367695777,411650080,1220644697]},{x:103758,v:[352823931,3335845976,4193218782]},{x:118934,m:v,n:[199,200],v:[3403189785]},{x:118711,v:[2364051618]},{x:81101,v:[1349146759,2206992415,2917249763,2946940121]},{x:57693,v:[2114570256,3129316290]},{x:107549,v:[1349146759]},{x:108677,v:[1444462398]},{x:123156,v:[2775654583]},{x:99441,v:[2866949877]},{x:112384,v:[1349146759,1444462398]},{x:123246,v:[2775654583]},{x:108616,v:[1444462398]},{x:122690,v:[2775654583]},{x:106973,v:[1349146759]},{x:122593,v:[1444462398]},{x:117361,v:[1349146759]},{x:85906,v:[2114570256]},{x:112379,v:[1444462398]},{x:123247,v:[2775654583]},{x:85901,v:[1783160115,3403189785]},{x:85902,v:[3928382683]},{x:85902,v:[3312905059,3928382683]},{x:85901,v:[1480368425,1783160115,3403189785]},{x:112379,v:[1349146759,1444462398]},{x:81100,v:[411650080,1220644697]},{x:103758,v:[352823931,3403189785,4193218782]},{x:103758,v:[3335845976]},{x:81100,v:[367695777]},{x:103758,v:[1349146759]},{x:118711,v:[1444462398]}]};function e(){var r,g,u={create:function(){for(var e=new Array(16),v=0;v<16;v++)e[v]=v%5==0?1:0;return e},perspective:function(e,v,n,t,r){v=1\/Math.tan(v\/2);return e[0]=v\/n,e[1]=0,e[2]=0,e[3]=0,e[4]=0,e[5]=v,e[6]=0,e[7]=0,e[8]=0,e[9]=0,e[11]=-1,e[12]=0,e[13]=0,e[15]=0,null!=r&&r!==1\/0?(e[10]=(r+t)*(n=1\/(t-r)),e[14]=2*r*t*n):(e[10]=-1,e[14]=-2*t),e},lookAt:function(e,v,n,t){var r,x,a,i,o=v[0],l=v[1],v=v[2],p=t[0],m=t[1],t=t[2],A=n[0],f=n[1],n=n[2];return Math.abs(o-A)<1e-6&&Math.abs(l-f)<1e-6&&Math.abs(v-n)<1e-6?u.identity(e):(A=o-A,f=l-f,n=v-n,r=m*(n*=i=1\/Math.hypot(A,f,n))-t*(f*=i),t=t*(A*=i)-p*n,p=p*f-m*A,(i=Math.hypot(r,t,p))?(r*=i=1\/i,t*=i,p*=i):p=t=r=0,m=f*p-n*t,x=n*r-A*p,a=A*t-f*r,(i=Math.hypot(m,x,a))?(m*=i=1\/i,x*=i,a*=i):a=x=m=0,e[0]=r,e[1]=m,e[2]=A,e[3]=0,e[4]=t,e[5]=x,e[6]=f,e[7]=0,e[8]=p,e[9]=a,e[10]=n,e[11]=0,e[12]=-(r*o+t*l+p*v),e[13]=-(m*o+x*l+a*v),e[14]=-(A*o+f*l+n*v),e[15]=1,e)},multiply:function(e,v,n){var t=v[0],r=v[1],x=v[2],a=v[3],i=v[4],o=v[5],l=v[6],p=v[7],m=v[8],A=v[9],f=v[10],u=v[11],c=v[12],h=v[13],s=v[14],v=v[15],d=n[0],P=n[1],g=n[2],U=n[3];return e[0]=d*t+P*i+g*m+U*c,e[1]=d*r+P*o+g*A+U*h,e[2]=d*x+P*l+g*f+U*s,e[3]=d*a+P*p+g*u+U*v,d=n[4],P=n[5],g=n[6],U=n[7],e[4]=d*t+P*i+g*m+U*c,e[5]=d*r+P*o+g*A+U*h,e[6]=d*x+P*l+g*f+U*s,e[7]=d*a+P*p+g*u+U*v,d=n[8],P=n[9],g=n[10],U=n[11],e[8]=d*t+P*i+g*m+U*c,e[9]=d*r+P*o+g*A+U*h,e[10]=d*x+P*l+g*f+U*s,e[11]=d*a+P*p+g*u+U*v,d=n[12],P=n[13],g=n[14],U=n[15],e[12]=d*t+P*i+g*m+U*c,e[13]=d*r+P*o+g*A+U*h,e[14]=d*x+P*l+g*f+U*s,e[15]=d*a+P*p+g*u+U*v,e},identity:function(e){return e[0]=1,e[1]=0,e[2]=0,e[3]=0,e[4]=0,e[5]=1,e[6]=0,e[7]=0,e[8]=0,e[9]=0,e[10]=1,e[11]=0,e[12]=0,e[13]=0,e[14]=0,e[15]=1,e}};function e(){var e,v,n,t;if(r=function(){x.width=67,x.height=67;var e=x.getContext(\"webgl\")||x.getContext(\"experimental-webgl\");e&&(e.viewport(0,0,67,67),e.clearColor(0,0,0,1),e.clear(e.COLOR_BUFFER_BIT));return e}())return e=r.createShader(r.VERTEX_SHADER),r.shaderSource(e,\"attribute vec3 c,d; uniform vec4 e; uniform vec3 f,g;uniform mat4 h,i;varying vec3 j;void main(){vec3 a=normalize(d);vec4 b=h*vec4(c,1.);vec3 k=normalize(vec3(e-b));j=g*f*max(dot(k,a),0.),gl_Position=i*vec4(c,1.);}\"),r.compileShader(e),n=r.createShader(r.FRAGMENT_SHADER),r.shaderSource(n,\"#ifdef GL_ES\\nprecision mediump float;\\n#endif\\nvarying vec3 j;void main(){gl_FragColor = vec4(j, 1.0);}\"),r.compileShader(n),g=r.createProgram(),r.attachShader(g,e),r.attachShader(g,n),r.linkProgram(g),r.detachShader(g,e),r.detachShader(g,n),r.deleteShader(e),r.deleteShader(n),r.useProgram(g),e=function(e){for(var v=[],n=[],t=[],r=[],x=0;x<=50;++x)for(var a=x*Math.PI\/50,i=Math.sin(a),o=Math.cos(a),l=0;l<=50;++l){var p=2*l*Math.PI\/50,m=Math.sin(p),p=Math.cos(p)*i,A=o,m=m*i,f=1-l\/50,u=1-x\/50;v.push(2*p),v.push(2*A),v.push(2*m),n.push(p),n.push(A),n.push(m),t.push(f),t.push(u)}for(x=0;x<50;++x)for(l=0;l<50;++l){var c=51*x+l,h=c+50+1;r.push(c),r.push(h),r.push(c+1),r.push(h),r.push(h+1),r.push(c+1)}var v=new Float32Array(v),n=new Float32Array(n),t=new Float32Array(t),r=new Uint16Array(r),s=e.createBuffer(),d=e.createBuffer(),P=e.createBuffer(),s=(e.bindBuffer(e.ARRAY_BUFFER,s),e.bufferData(e.ARRAY_BUFFER,v,e.STATIC_DRAW),e.getAttribLocation(g,\"c\")),s=(e.vertexAttribPointer(s,3,e.FLOAT,!1,0,0),e.enableVertexAttribArray(s),e.bindBuffer(e.ARRAY_BUFFER,d),e.bufferData(e.ARRAY_BUFFER,n,e.STATIC_DRAW),e.getAttribLocation(g,\"d\"));return e.vertexAttribPointer(s,3,e.FLOAT,!1,0,0),e.enableVertexAttribArray(s),e.bindBuffer(e.ELEMENT_ARRAY_BUFFER,P),e.bufferData(e.ELEMENT_ARRAY_BUFFER,r,e.STATIC_DRAW),r.length}(r),r.clearColor(0,0,0,1),r.enable(r.DEPTH_TEST),n=u.create(),u.perspective(n,Math.PI\/6,1,.1,100),v=u.create(),u.lookAt(v,[0,0,10],[0,0,0],[0,1,0]),t=u.create(),u.multiply(t,n,v),n=r.getUniformLocation(g,\"h\"),r.uniformMatrix4fv(n,!1,v),n=r.getUniformLocation(g,\"i\"),r.uniformMatrix4fv(n,!1,t),v=r.getUniformLocation(g,\"e\"),r.uniform4fv(v,[10,10,10,1]),n=r.getUniformLocation(g,\"f\"),r.uniform3fv(n,[.9,.5,.3]),t=r.getUniformLocation(g,\"g\"),r.uniform3fv(t,[1,1,1]),r.clear(r.COLOR_BUFFER_BIT|r.DEPTH_BUFFER_BIT),r.drawElements(r.TRIANGLES,e,r.UNSIGNED_SHORT,0),r.useProgram(null),g&&r.deleteProgram(g),x.toDataURL()}var v,x,n=0;return n=null!=(x=document.createElement(\"canvas\"))&&(v=e())?t(v):n}function t(e){for(var v=2166136261,n=0;n<e.length;++n)v=(v^=e.charCodeAt(n))+((v<<1)+(v<<4)+(v<<7)+(v<<8)+(v<<24));return v>>>0}function v(){var e=0,v=document.createElement(\"canvas\");return e=null!=v&&(v=function(e){e.width=67,e.height=67;var v=e.getContext(\"2d\",{alpha:!0});if(null!=v)return v.imageSmoothingQuality=\"low\",v.imageSmoothingEnabled=!0,v.globalCompositeOperation=\"source-over\",v.globalAlpha=1,v.miterLimit=1\/0,v.filter=\"none\",v.lineCap=\"butt\",v.lineDashOffset=0,v.lineJoin=\"miter\",v.font=\"10pt Arial\",v.lineWidth=2,void 0!==v.setLineDash&&v.setLineDash([10,20]),v.shadowColor=\"black\",v.shadowOffsetX=-3,v.shadowOffsetY=-5,v.translate(e.width\/2,e.height\/2),v.rotate(.8901179),v.fillStyle=\"green\",v.textAlign=\"center\",v.textBaseline=\"middle\",v.fillText(\"*51Degrees*\",0,0),v.beginPath(),v.shadowColor=\"yellow\",v.shadowBlur=1,v.shadowOffsetX=1,v.shadowOffsetY=1,v.strokeStyle=\"red\",v.fillStyle=\"rgba(0, 0, 255, 0.6)\",void 0===v.ellipse?v.arc(0,0,25,0,2*Math.PI):v.ellipse(0,0,25,15,Math.PI\/4,0,2*Math.PI),v.fill(),v.stroke(),e.toDataURL()}(v))?t(v):e}function n(){var e=document.createElement(\"canvas\");if(null!=e){e=e.getContext(\"webgl\")||e.getContext(\"experimental-webgl\");if(e){var v=e.getExtension(\"WEBGL_debug_renderer_info\");if(v)return e.getParameter(v.UNMASKED_RENDERER_WEBGL)}}return\"\"}function r(e,v){for(var n,t=0;t<v.length;t++)if(n=\"(\"+e+\": \"+v[t]+\")\",window.matchMedia(n).matches)return v[t];return\"n\/a\"}function x(){return r(\"color-gamut\",[\"p3\",\"srgb\"])}function f(e,v){if(e.m){var n=e.m(e);if(n||\"\"===n){if(!n.then){for(var t=e,r=n,x=v,a=0;a<t.n.length;a++){var i=A.Data[t.n[a]];if(i.r)for(var o=0;o<i.r.length;o++){var l=i.r[o];if((null===l.a||r>=l.a)&&(null===l.b||r<=l.b))return void f(i,0)}else if(i.v&&-1!=i.v.indexOf(r))return void f(i,0)}0<t.n.length&&x<m?setTimeout(function(){f(t,x+1)},10):p(t.x)}}else e.x&&p(e.x)}else p(e.x)}f(A.Data[0],0)}getProfileId(function(e){profileIds.push(e),document.cookie=\"51D_ProfileIds=\"+profileIds.join(\"|\")});\ndocument.cookie = \"51D_ProfileIds=\" + profileIds.join(\"|\")","javascriptbandwidth":"function FODBW() {\r\n\r\n    this.getCookie = function (name) {\r\n        var value = document.cookie;\r\n        var start = value.indexOf(\" \" + name + \"=\");\r\n        if (start == -1) {\r\n            start = value.indexOf(name + \"=\");\r\n        }\r\n        if (start == -1) {\r\n            value = null;\r\n        }\r\n        else {\r\n            start = value.indexOf(\"=\", start) + 1;\r\n            var end = value.indexOf(\";\", start);\r\n            if (end == -1) {\r\n                end = value.length;\r\n            }\r\n            value = decodeURIComponent(value.substring(start,end));\r\n        }\r\n        return value;\r\n    }\r\n\r\n    this.unloadBegin = function () {\r\n        if (this.unloadComplete == false &&\r\n            this.id != undefined &&\r\n            this.loadStartTime != undefined &&\r\n            this.loadCompleteTime != undefined &&\r\n            this.pageLength != undefined) {\r\n            var value = this.id + \"|\" + this.loadStartTime + \"|\" + new Date().getTime() + \"|\" + this.loadCompleteTime + \"|\" + this.pageLength;\r\n            document.cookie = \"51D_Bandwidth\" + \"=\" + encodeURIComponent(value) + \"; path=\/\";\r\n            this.unloadComplete = true;\r\n        }\r\n    }\r\n\r\n    this.loadComplete = function () {\r\n        if (this.loadCompleteSet == false) {\r\n            this.pageLength = document.body.innerHTML.length + document.head.innerHTML.length;\r\n            this.loadCompleteTime = new Date().getTime();\r\n            this.loadCompleteSet = true;\r\n        }\r\n    }\r\n\r\n    this.loadStartTime = new Date().getTime();\r\n    this.id = this.getCookie(\"51D_Bandwidth\");\r\n    this.unloadComplete = false;\r\n    this.loadCompleteSet = false;\r\n\r\n   if (window.addEventListener) {\r\n                window.addEventListener(\"beforeunload\", this.unloadBegin.bind(this), true);\r\n                window.addEventListener(\"click\", this.unloadBegin.bind(this), true);\r\n                if (document.readyState === \"complete\") {\r\n                                                this.pageLength = document.body.innerHTML.length +    document.head.innerHTML.length;\r\n                                                this.loadCompleteTime = new Date().getTime();\r\n                }\r\n                document.addEventListener(\"load\", this.loadComplete.bind(this), true);\r\n                document.addEventListener(\"DOMContentLoaded\", this.loadComplete.bind(this), true);\r\n   } else {\r\n                window.attachEvent(\"beforeunload\", this.unloadBegin.bind(this), true);\r\n                window.attachEvent(\"click\", this.unloadBegin.bind(this), true);\r\n                if (document.readyState === \"complete\") {\r\n                                                this.pageLength = document.body.innerHTML.length +    document.head.innerHTML.length;\r\n                                                this.loadCompleteTime = new Date().getTime();\r\n                }\r\n                document.attachEvent(\"load\", this.loadComplete.bind(this), true);\r\n                document.attachEvent(\"DOMContentLoaded\", this.loadComplete.bind(this), true);\r\n   }\r\n}\r\n","javascriptimageoptimiser":"function FODIO(width_param, height_param) {\r\n\r\n    var src_parameter = \"src\";\r\n    var width_parameter = \"w\";\r\n    if (typeof width_param !== \"undefined\") { width_parameter = width_param; }\r\n    var height_parameter = \"h\";\r\n    if (typeof height_param !== \"undefined\") { height_parameter = height_param; }\r\n    var auto_parameter = \"auto\";\r\n\r\n    \/\/* (LoadImage) Loads the image based on measurements taken within the browser. *\/\r\n    function setImage(image_element, image_src) {\r\n\r\n        \/\/ Get the query string parameters for the image element src.\r\n        var qs = (function (a) {\r\n            var b = {};\r\n            if (a != \"\") {\r\n                for (var i = 0; i < a.length; ++i) {\r\n                    var p = a[i].split('=');\r\n                    if (p.length != 2) continue;\r\n                    b[p[0]] = decodeURIComponent(p[1].replace(\/\\+\/g, \" \"));\r\n                }\r\n            }\r\n            return b;\r\n        })(image_src.split(\/[&\\?]+\/));\r\n\r\n        \/\/ Does the control have a 1 by 1 image displayed?\r\n        var startWithParent = (image_element.parentNode.nodeName == \"A\" ||\r\n        (image_element.offsetWidth == 1 &&\r\n        image_element.offsetHeight == 1));\r\n\r\n        \/\/ Get the width of the image.\r\n        var width = qs[width_parameter];\r\n        if (width != undefined && isNaN(width)) {\r\n            width = getWidth(startWithParent ? image_element.parentNode : image_element);\r\n            if (startWithParent)\r\n                image_element.parentNode.style.width = width + \"px\";\r\n        }\r\n\r\n        \/\/ Get the height of the image.\r\n        var height = qs[height_parameter];\r\n        if (height != undefined && isNaN(height)) {\r\n            height = getHeight(startWithParent ? image_element.parentNode : image_element);\r\n            if (startWithParent)\r\n                image_element.parentNode.style.height = height + \"px\";\r\n        }\r\n\r\n        \/\/ Change the image source to the selected width and height.\r\n        var parameters = [];\r\n\r\n        if (width > 0)\r\n            parameters.push(width_parameter + \"=\" + width);\r\n        if (height > 0)\r\n            parameters.push(height_parameter + \"=\" + height);\r\n\t\tif(qs[src_parameter] != undefined)\r\n\t\t\tparameters.push(src_parameter + \"=\" + qs[src_parameter]);\r\n\r\n        image_element.src = image_src.substring(0, image_src.indexOf(\"?\")) +\r\n        \"?\" + parameters.join(\"&\");\r\n    }\r\n\r\n    function getWidth(element) {\r\n        if (element != undefined) {\r\n            if (element.nodeName != \"A\" && element.offsetWidth != 0)\r\n                return element.offsetWidth;\r\n            return getWidth(element.parentNode);\r\n        }\r\n        return 0;\r\n    }\r\n\r\n    function getHeight(element) {\r\n        if (element != undefined) {\r\n            if (element.nodeName != \"A\" && element.offsetHeight != 0)\r\n                return element.offsetHeight;\r\n            return getHeight(element.parentNode);\r\n        }\r\n        return 0;\r\n    }\r\n\r\n    function setImageSource(image_element, image_src) {\r\n        if (image_src.indexOf(\"?\") >= 0 && (\r\n            image_src.indexOf(width_parameter + \"=\" + auto_parameter, image_src.indexOf(\"?\")) >= 0 ||\r\n            image_src.indexOf(height_parameter + \"=\" + auto_parameter, image_src.indexOf(\"?\")) >= 0)) {\r\n            setImage(image_element, image_src);\r\n            return true;\r\n        }\r\n        return false;\r\n    }\r\n\r\n    \/\/ Loop through all the img elements finding any with the data src\r\n    \/\/ attribute set.\r\n    var images = document.getElementsByTagName(\"img\");\r\n    for (var i = 0, max = images.length; i < max; i++) {\r\n        var image_element = images[i];\r\n        if (!setImageSource(image_element, image_element.src)) {\r\n            var data_src = image_element.attributes[\"data-src\"];\r\n            if (data_src != undefined) {\r\n                setImageSource(image_element, data_src.value);\r\n            }\r\n        }\r\n    }\r\n}","browseraudiocodecsdecode":["Unknown"],"browservideocodecsdecode":["Unknown"],"priceband":"Unknown","screenmmsquare":0,"screenmmsquarenullreason":"Unknown","screenmmdiagonalrounded":0,"screenmmdiagonalroundednullreason":"Unknown","screeninchessquare":0,"screeninchessquarenullreason":"Unknown","screeninchesdiagonalrounded":0,"screeninchesdiagonalroundednullreason":"Unknown","devicetype":"Desktop","browserpropertysource":"NotSet","supportswebgl":false,"supportswebglnullreason":"Unknown","isemulatingdesktop":false,"isemulatingdesktopnullreason":"Unknown","supportsphonecalls":false,"supportsphonecallsnullreason":"Unknown","iswebapp":false,"iswebappnullreason":"Unknown","meter":false,"meternullreason":"Unknown","issmartwatch":false,"issmartwatchnullreason":"Unknown","nativeplatform":"N\/A","nativebrand":["N\/A"],"nativedevice":["N\/A"],"nativemodel":["N\/A"],"nativename":["N\/A"],"screenpixelsheightjavascript":"","screenpixelsheightjavascriptnullreason":"Unknown","screenpixelswidthjavascript":"","screenpixelswidthjavascriptnullreason":"Unknown","devicecertifications":["Unknown"],"secondbackcameramegapixels":0,"secondbackcameramegapixelsnullreason":"Unknown","hardwareaudiocodecsdecode":["Unknown"],"hardwarevideocodecsdecode":["Unknown"],"softwareaudiocodecsdecode":["AAC","ADPCM","ALAC","PCM","QCELP"],"softwareaudiocodecsencode":["ALAC"],"softwarevideocodecsdecode":["Unknown"],"softwarevideocodecsencode":["Unknown"],"platformreleaseyear":0,"platformreleaseyearnullreason":"Unknown","platformreleasemonth":"Unknown","platformpreviewmonth":"May","platformpreviewyear":2023,"platformdiscontinuedmonth":"N\/A","platformdiscontinuedyear":0,"platformdiscontinuedyearnullreason":"Unknown","browserpreviewmonth":"May","browserpreviewyear":2023,"browserreleasemonth":"Unknown","browserreleaseyear":0,"browserreleaseyearnullreason":"Unknown","browserdiscontinuedmonth":"N\/A","browserdiscontinuedyear":0,"browserdiscontinuedyearnullreason":"Unknown","hardwareaudiocodecsencode":["Unknown"],"hardwarevideocodecsencode":["Unknown"],"isemulatingdevice":false,"isemulatingdevicenullreason":"Unknown","isdataminimising":false,"isdataminimisingnullreason":"Unknown","secondfrontcameramegapixels":0,"secondfrontcameramegapixelsnullreason":"Unknown","browseraudiocodecsencode":["Unknown"],"browservideocodecsencode":["Unknown"],"specificabsorbtionrateus":0,"specificabsorbtionrateusnullreason":"Unknown","specificabsorbtionrateeu":0,"specificabsorbtionrateeunullreason":"Unknown","weightwithbattery":0,"weightwithbatterynullreason":"Unknown","weightwithoutbattery":0,"weightwithoutbatterynullreason":"Unknown","httplivestreaming":"False","hardwaremodelvariants":["N\/A"],"hardwarecarrier":"N\/A","hardwarerank":4,"platformrank":821,"platformpreviewage":2,"platformreleaseage":0,"platformreleaseagenullreason":"Unknown","platformdiscontinuedage":0,"platformdiscontinuedagenullreason":"Unknown","browserrank":3080,"browserpreviewage":2,"browserreleaseage":0,"browserreleaseagenullreason":"Unknown","browserdiscontinuedage":0,"browserdiscontinuedagenullreason":"Unknown","releaseage":0,"releaseagenullreason":"Unknown","ishardwaregroup":false,"ishardwaregroupnullreason":"Unknown","internalstoragevariants":["Unknown"],"promise":"None","hardwareprofilesource":"N\/A","deviceramvariants":["Unknown"],"frequencybands":["Unknown"],"invrmode":false,"invrmodenullreason":"Unknown","isscreenfoldable":false,"isscreenfoldablenullreason":"Unknown","secondscreeninchesdiagonal":0,"secondscreeninchesdiagonalnullreason":"Unknown","secondscreenincheswidth":0,"secondscreenincheswidthnullreason":"Unknown","secondscreenpixelswidth":0,"secondscreenpixelswidthnullreason":"Unknown","secondscreeninchesdiagonalrounded":0,"secondscreeninchesdiagonalroundednullreason":"Unknown","secondscreeninchessquare":0,"secondscreeninchessquarenullreason":"Unknown","secondscreeninchesheight":0,"secondscreeninchesheightnullreason":"Unknown","secondscreenmmdiagonal":0,"secondscreenmmdiagonalnullreason":"Unknown","secondscreenmmdiagonalrounded":0,"secondscreenmmdiagonalroundednullreason":"Unknown","secondscreenmmheight":0,"secondscreenmmheightnullreason":"Unknown","secondscreenmmsquare":0,"secondscreenmmsquarenullreason":"Unknown","secondscreenmmwidth":0,"secondscreenmmwidthnullreason":"Unknown","secondscreenpixelsheight":0,"secondscreenpixelsheightnullreason":"Unknown","tac":["Unknown"],"cssgrid":false,"cssgridnullreason":"Unknown","fetch":false,"fetchnullreason":"Unknown","webp":true,"numberofscreens":0,"numberofscreensnullreason":"Unknown","http2":false,"http2nullreason":"Unknown","preload":false,"preloadnullreason":"Unknown","jpeg2000":false,"jpeg2000nullreason":"Unknown","browserfamily":"Edge","pixelratio":1,"pixelratiojavascript":"","pixelratiojavascriptnullreason":"Unknown","setheaderbrowseraccept-ch":"Sec-CH-UA,Sec-CH-UA-Full-Version-List,Sec-CH-UA-Mobile,Sec-CH-UA-Platform","setheaderhardwareaccept-ch":"Sec-CH-UA-Model,Sec-CH-UA-Mobile","setheaderplatformaccept-ch":"Sec-CH-UA-Platform,Sec-CH-UA-Platform-Version","javascriptgethighentropyvalues":"if(navigator.userAgentData){navigator.userAgentData.getHighEntropyValues([\"model\",\"platform\",\"platformVersion\",\"fullVersionList\"]).then(t=>{document.cookie=`51D_GetHighEntropyValues=${btoa(JSON.stringify(t))}`\r\n\/\/ 51D replace this comment with callback function.\r\n})} else { \/\/ 51D replace this comment with callback function.\r\n}","matchednodes":18,"difference":0,"differencenullreason":"Unknown","drift":0,"driftnullreason":"Unknown","deviceid":"17595-128124-126072-18092","useragents":["?0","_macOS\"","________________intosh________________ 10_15_7) App______________6 (KHT____________________me\/115.0.0_0 Safa_________","__acOS\"\u001f_____","_______\u001f\"13.5___","____________________________________________v=\"115__________________________________________\"\u001f?0\u001f\"macOS_","____________\";v=\"9_________________________5\", \"C____________________\"macOS_"],"iterations":75,"method":0,"methodnullreason":"Unknown"}}
```

Thus it is important to 
a) fix the warnings
b) suppress this output

I'll create a separate issue

**Comment by justadreamer** [comment]
according to @drasmart research: 

### App killed before json sent

```text
POST http://localhost:5101/51dpipeline/json net::ERR_CONNECTION_REFUSED
```

### Offline via dev tools

```text
POST http://localhost:5101/51dpipeline/json net::ERR_INTERNET_DISCONNECTED
```

### Timeout

_(There is no timeout set on XMLHttpRequest.)_

### CORS block

```text
Access to XMLHttpRequest at 'https://localhost:5001/51dpipeline/json' from origin 'http://localhost:5101' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header has a value 'https://localhost:5301' that is not equal to the supplied origin.
```

### (Effectively) Empty JSON

```text
Uncaught TypeError: Cannot read properties of undefined (reading 'hardwarename')
    at (index):246:51
    at Array.<anonymous> (51Degrees.core.js:1:5478)
    at y (51Degrees.core.js:1:4363)
    at p (51Degrees.core.js:1:4427)
    at u.onload (51Degrees.core.js:1:4999)
```

## Suggested amendments to the error message
So in the case json was not parsable - the error should be caught and logged into the console: "/json endpoint did not return a valid JSON - please verify the setup and CORS headers for the endpoint".  the fod.complete still should be called with an empty JSON.
